core/src/main/java/org/apache/spark/io/ReadAheadInputStream.java

```diff
@@ -0,0 +1,289 @@
+      stateChangeLock.unlock();
+    }
+  }
+
+  private void waitForAsyncReadComplete() {
+    stateChangeLock.lock();
+    try {
+      asyncReadComplete.await();
+    } catch (InterruptedException e) {
+    } finally {
+      stateChangeLock.unlock();
+    }
+  }
+
+  @Override
+  public synchronized int read() throws IOException {
+    int val = read(oneByte, 0, 1);
+    if (val == -1) {
+      return -1;
+    }
+    return oneByte[0] & 0xFF;
+  }
+
+  @Override
+  public synchronized int read(byte[] b, int offset, int len) throws IOException {
+    stateChangeLock.lock();
+    try {
+      len = readInternal(b, offset, len);
+    }
+    finally {
+      stateChangeLock.unlock();
+    }
+    return len;
+  }
+
+  /**
+   * Internal read function which should be called only from read() api. The assumption is that
+   * the stateChangeLock is already acquired in the caller before calling this function.
+   */
+  private int readInternal(byte[] b, int offset, int len) throws IOException {
+    assert (stateChangeLock.isLocked());
+    if (offset < 0 || len < 0 || offset + len < 0 || offset + len > b.length) {
+      throw new IndexOutOfBoundsException();
+    }
+    if (!activeBuffer.hasRemaining()) {
+      if (!isReadInProgress) {
+        // This condition will only be triggered for the first time read is called.
+        readAsync(activeBuffer);
+      }
+      waitForAsyncReadComplete();
+    }
+
+    if (isReadAborted) {
+      throw new IOException(readException);
+    }
+    if (isEndOfStream()) {
+      return -1;
+    }
+    len = Math.min(len, activeBuffer.remaining());
+    activeBuffer.get(b, offset, len);
+
+    if (activeBuffer.remaining() <= readAheadThresholdInBytes && !readAheadBuffer.hasRemaining()) {
+      readAsync(readAheadBuffer);
+    }
+    if (!activeBuffer.hasRemaining()) {
+      ByteBuffer temp = activeBuffer;
+      activeBuffer = readAheadBuffer;
+      readAheadBuffer = temp;
+    }
+    return len;
+  }
+
+  @Override
+  public synchronized int available() throws IOException {
+    stateChangeLock.lock();
+    int val = activeBuffer.remaining() + readAheadBuffer.remaining();
+    stateChangeLock.unlock();
+    return val;
+  }
+
+  @Override
+  public synchronized long skip(long n) throws IOException {
+    stateChangeLock.lock();
+    long skipped;
+    try {
+      skipped = skipInternal(n);
+    } finally {
+      stateChangeLock.unlock();
+    }
+    return skipped;
+  }
+
+  /**
+   * Internal skip function which should be called only from skip() api. The assumption is that
+   * the stateChangeLock is already acquired in the caller before calling this function.
+   */
+  private long skipInternal(long n) throws IOException {
+    assert (stateChangeLock.isLocked());
+    if (n <= 0L) {
+      return 0L;
+    }
+    if (isReadInProgress) {
+      waitForAsyncReadComplete();
+    }
+    if (available() >= n) {
+      // we can skip from the internal buffers
+      int toSkip = Ints.checkedCast(n);
```

Comment:
In general, checkedCast must be used when parameter must be an Int as a precondition [1] - not when it is being coerced into an int as part of impl detail.
It is valid for n to be > Integer.MAX_VALUE as a parameter value.

Having said that : available() is an int - so if n is less than available, it is within Int bounds.
Using Ints.checkedCast here would be incorrect from correctness and logical point of view.

[1] checkedCast will throw an exception - which would have been the wrong behavior here given contract says n is long.